src/zhttp_server.c

```diff
@@ -0,0 +1,512 @@
+/*  =========================================================================
+    zhttp_server -
+
+    Copyright (c) the Contributors as noted in the AUTHORS file.
+
+    This Source Code Form is subject to the terms of the Mozilla Public
+    License, v. 2.0. If a copy of the MPL was not distributed with this
+    file, You can obtain one at http://mozilla.org/MPL/2.0/.
+    =========================================================================
+*/
+
+/*
+@header
+    zhttp_server -
+@discuss
+@end
+*/
+
+#include "czmq_classes.h"
+
+#ifdef HAVE_LIBMICROHTTPD
+#include <microhttpd.h>
+#endif
+
+#ifdef HAVE_LIBMICROHTTPD
+
+typedef struct {
+    char* content;
+    size_t content_length;
+    zhash_t *headers;
+} request_t;
+
+typedef struct {
+    zsock_t *pipe;              //  Actor command pipe
+    bool terminated;            //  Did caller ask us to quit?
+    bool verbose;               //  Verbose logging enabled?
+
+    zhttp_server_options_t *options;
+    zsock_t *backend;
+    struct MHD_Daemon *daemon;
+    zframe_t *empty_frame;
+    zlistx_t* pending_connections;
+} server_actor_t;
+
+static int
+s_handle_request (
+        server_actor_t *self,
+        struct MHD_Connection *mhd_connection,
+        const char *url,
+        const char *method,
+        const char *version,
+        const char *upload_data,
+        size_t *upload_data_size,
+        request_t **request_p);
+
+static void
+s_handle_request_completed (
+        server_actor_t *self,
+        struct MHD_Connection *mhd_connection,
+        request_t **request_p,
+        enum MHD_RequestTerminationCode toe);
+
+static void
+server_actor (zsock_t *pipe, void *args);
+
+
+static request_t*
+request_new () {
+    request_t *self = (request_t*)zmalloc (sizeof (request_t));
+    assert (self);
+
+    self->headers = zhash_new ();
+    zhash_autofree (self->headers);
+
+    return self;
+}
+
+static void
+request_destroy (request_t **self_p) {
+    request_t * self= *self_p;
+
+    if (self) {
+        zhash_destroy (&self->headers);
+        zstr_free (&self->content);
+
+        free (self);
+    }
+
+    *self_p = NULL;
+}
+
+static void *
+s_insert_connection (struct MHD_Connection *connection) {
+    MHD_suspend_connection (connection);
```

Comment:
this API and the next were also introduced in 0.9.34: https://github.com/scottjg/libmicrohttpd/commit/64a2c30ad3c3476bf16f0b2b579e040c6b65922a is there an alternative?
I can backport for the OBS projects if needed

Reply:
@somdoron

Reply:
I'm not sure it will work...
Actually I think it will work, but will cause very high CPU as microhttpd will continue call the handler function until response is provided.

Reply:
I think that as long a we document that, it's fine. It's an optional feature and we can recommend to use at least 0.9.34 if the intention is to deploy in production. What do you think?

Reply:
Sounds good to me. 
I'm not 100% sure it will work, but I think so.
So just skip the suspend and resume functions if not available.

Reply:
Ok, will send a PR later